File: packages/vm/wasmlib/ts/wasmlib/bytes.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {Convert} from "./convert";
import {ScAddress, ScAgentID, ScChainID, ScColor, ScHash, ScHname, ScRequestID} from "./hashtypes";
import {panic} from "./host";

// decodes separate entities from a byte buffer
export class BytesDecoder {
    buf: u8[];

    // constructs a decoder
    constructor(data: u8[]) {
        if (data.length == 0) {
            panic("cannot decode empty byte array, use exist()");
        }
        this.buf = data;
    }

    // decodes an ScAddress from the byte buffer
    address(): ScAddress {
        return ScAddress.fromBytes(this.bytes());
    }

    // decodes an ScAgentID from the byte buffer
    agentID(): ScAgentID {
        return ScAgentID.fromBytes(this.bytes());
    }

    // decodes a bool from the byte buffer
    bool(): boolean {
        return this.uint8() != 0;
    }

    // decodes the next substring of bytes from the byte buffer
    bytes(): u8[] {
        let size = this.uint32();
        if (u32(this.buf.length) < size) {
            panic("insufficient bytes");
        }
        let value = this.buf.slice(0, size);
        this.buf = this.buf.slice(size);
        return value;
    }

    // decodes an ScChainID from the byte buffer
    chainID(): ScChainID {
        return ScChainID.fromBytes(this.bytes());
    }

    // decodes an ScColor from the byte buffer
    color(): ScColor {
        return ScColor.fromBytes(this.bytes());
    }

    // decodes an ScHash from the byte buffer
    hash(): ScHash {
        return ScHash.fromBytes(this.bytes());
    }

    // decodes an ScHname from the byte buffer
    hname(): ScHname {
        return ScHname.fromBytes(this.bytes());
    }

    // decodes an int8 from the byte buffer
    int8(): i8 {
        return this.uint8() as i8;
    }

    // decodes an int16 from the byte buffer
    // note that these are encoded using leb128 encoding to conserve space
    int16(): i16 {
        return this.leb128Decode(16) as i16;
    }

    // decodes an int32 from the byte buffer
    // note that these are encoded using leb128 encoding to conserve space
    int32(): i32 {
        return this.leb128Decode(32) as i32;
    }

    // decodes an int64 from the byte buffer
    // note that these are encoded using leb128 encoding to conserve space
    int64(): i64 {
        return this.leb128Decode(64);
    }

    // leb128 decoder
    leb128Decode(bits: i32): i64 {
        let val: i64 = 0;
        let s = 0;
        for (; ;) {
            if (this.buf.length == 0) {
                panic("leb128Decode: insufficient bytes");
            }
            let b = this.buf.shift() as i8;
            val |= ((b & 0x7f) as i64) << s;

            // termination bit set?
            if ((b & -0x80) == 0) {
                if ((((val >> s) as i8) & 0x7f) != (b & 0x7f)) {
                    panic("integer too large");
                }

                // extend int7 sign to int8
                b |= (b & 0x40) << 1;

                // extend int8 sign to int64
                val |= ((b as i64) << s);
                break;
            }
            s += 7;
            if (s >= bits) {
                panic("integer representation too long");
            }
        }
        return val;
    }

    // decodes an ScRequestID from the byte buffer
    requestID(): ScRequestID {
        return ScRequestID.fromBytes(this.bytes());
    }

    // decodes an UTF-8 text string from the byte buffer
    string(): string {
        return Convert.toString(this.bytes());
    }

    // decodes an uint8 from the byte buffer
    uint8(): u8 {
        return this.buf.shift();
    }

    // decodes an uint16 from the byte buffer
    // note that these are encoded using leb128 encoding to conserve space
    uint16(): u16 {
        return this.int16() as u16;
    }

    // decodes an uint32 from the byte buffer
    // note that these are encoded using leb128 encoding to conserve space
    uint32(): u32 {
        return this.int32() as u32;
    }

    // decodes an uint64 from the byte buffer
    // note that these are encoded using leb128 encoding to conserve space
    uint64(): u64 {
        return this.int64() as u64;
    }

    close(): void {
        if (this.buf.length != 0) {
            panic("extra bytes");
        }
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// encodes separate entities into a byte buffer
export class BytesEncoder {
    buf: u8[];

    // constructs an encoder
    constructor() {
        this.buf = []
    }

    // encodes an ScAddress into the byte buffer
    address(value: ScAddress): BytesEncoder {
        return this.bytes(value.toBytes());
    }

    // encodes an ScAgentID into the byte buffer
    agentID(value: ScAgentID): BytesEncoder {
        return this.bytes(value.toBytes());
    }

    // encodes a bool into the byte buffer
    bool(val: boolean): BytesEncoder {
         return this.int8(val ? 1 : 0);
    }

    // encodes a substring of bytes into the byte buffer
    bytes(value: u8[]): BytesEncoder {
        this.uint32(value.length);
        for (let i = 0; i < value.length; i++) {
            this.buf.push(value[i]);
        }
        return this;
    }

    // encodes an ScChainID into the byte buffer
    chainID(value: ScChainID): BytesEncoder {
        return this.bytes(value.toBytes());
    }

    // encodes an ScColor into the byte buffer
    color(value: ScColor): BytesEncoder {
        return this.bytes(value.toBytes());
    }

    // retrieve the encoded byte buffer
    data(): u8[] {
        return this.buf;
    }

    // encodes an ScHash into the byte buffer
    hash(value: ScHash): BytesEncoder {
        return this.bytes(value.toBytes());
    }

    // encodes an ScHname into the byte buffer
    hname(value: ScHname): BytesEncoder {
        return this.bytes(value.toBytes());
    }

    // encodes an int8 into the byte buffer
    int8(val: i8): BytesEncoder {
        return this.uint8(val as u8);
    }

    // encodes an int16 into the byte buffer
    // note that these are encoded using leb128 encoding to conserve space
    int16(val: i16): BytesEncoder {
        return this.leb128Encode(val as i64);
    }

    // encodes an int32 into the byte buffer
    // note that these are encoded using leb128 encoding to conserve space
    int32(val: i32): BytesEncoder {
        return this.leb128Encode(val as i64);
    }

    // encodes an int64 into the byte buffer
    // note that these are encoded using leb128 encoding to conserve space
    int64(val: i64): BytesEncoder {
        return this.leb128Encode(val);
    }

    // leb128 encoder
    leb128Encode(val: i64): BytesEncoder {
        for (; ;) {
            let b = val as u8;
            let s = b & 0x40;
            val >>= 7;
            if ((val == 0 && s == 0) || (val == -1 && s != 0)) {
                this.buf.push(b & 0x7f);
                break;
            }
            this.buf.push(b | 0x80);
        }
        return this;
    }

    // encodes an ScRequestID into the byte buffer
    requestID(value: ScRequestID): BytesEncoder {
        return this.bytes(value.toBytes());
    }

    // encodes an UTF-8 text string into the byte buffer
    string(value: string): BytesEncoder {
        return this.bytes(Convert.fromString(value));
    }

    // encodes an uint8 into the byte buffer
    uint8(val: u8): BytesEncoder {
        this.buf.push(val);
        return this;
    }

    // encodes an uint16 into the byte buffer
    // note that these are encoded using leb128 encoding to conserve space
    uint16(val: u16): BytesEncoder {
        return this.int16(val as i16);
    }

    // encodes an uint32 into the byte buffer
    // note that these are encoded using leb128 encoding to conserve space
    uint32(val: u32): BytesEncoder {
        return this.int32(val as i32);
    }

    // encodes an uint64 into the byte buffer
    // note that these are encoded using leb128 encoding to conserve space
    uint64(val: u64): BytesEncoder {
        return this.int64(val as i64);
    }
}

File: packages/vm/wasmlib/go/wasmlib/bytes.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmlib

type BytesDecoder struct {
	data []byte
}

func NewBytesDecoder(data []byte) *BytesDecoder {
	if len(data) == 0 {
		panic("cannot decode empty byte array, use exist()")
	}
	return &BytesDecoder{data: data}
}

func (d *BytesDecoder) Address() ScAddress {
	return NewScAddressFromBytes(d.FixedBytes(uint32(TypeSizes[TYPE_ADDRESS])))
}

func (d *BytesDecoder) AgentID() ScAgentID {
	return NewScAgentIDFromBytes(d.FixedBytes(uint32(TypeSizes[TYPE_AGENT_ID])))
}

func (d *BytesDecoder) Bool() bool {
	return d.Uint8() != 0
}

func (d *BytesDecoder) Bytes() []byte {
	length := d.Uint32()
	return d.FixedBytes(length)
}

func (d *BytesDecoder) ChainID() ScChainID {
	return NewScChainIDFromBytes(d.FixedBytes(uint32(TypeSizes[TYPE_CHAIN_ID])))
}

func (d *BytesDecoder) Close() {
	if len(d.data) != 0 {
		panic("extra bytes")
	}
}

func (d *BytesDecoder) Color() ScColor {
	return NewScColorFromBytes(d.FixedBytes(uint32(TypeSizes[TYPE_COLOR])))
}

func (d *BytesDecoder) FixedBytes(size uint32) []byte {
	if uint32(len(d.data)) < size {
		panic("insufficient bytes")
	}
	value := d.data[:size]
	d.data = d.data[size:]
	return value
}

func (d *BytesDecoder) Hash() ScHash {
	return NewScHashFromBytes(d.FixedBytes(uint32(TypeSizes[TYPE_HASH])))
}

func (d *BytesDecoder) Hname() ScHname {
	return NewScHnameFromBytes(d.FixedBytes(uint32(TypeSizes[TYPE_HNAME])))
}

func (d *BytesDecoder) Int8() int8 {
	return int8(d.Uint8())
}

func (d *BytesDecoder) Int16() int16 {
	return int16(d.vliDecode(16))
}

func (d *BytesDecoder) Int32() int32 {
	return int32(d.vliDecode(32))
}

func (d *BytesDecoder) Int64() int64 {
	return d.vliDecode(64)
}

func (d *BytesDecoder) RequestID() ScRequestID {
	return NewScRequestIDFromBytes(d.FixedBytes(uint32(TypeSizes[TYPE_REQUEST_ID])))
}

func (d *BytesDecoder) String() string {
	return string(d.Bytes())
}

func (d *BytesDecoder) Uint8() uint8 {
	if len(d.data) == 0 {
		panic("insufficient bytes")
	}
	value := d.data[0]
	d.data = d.data[1:]
	return value
}

func (d *BytesDecoder) Uint16() uint16 {
	return uint16(d.vluDecode(16))
}

func (d *BytesDecoder) Uint32() uint32 {
	return uint32(d.vluDecode(32))
}

func (d *BytesDecoder) Uint64() uint64 {
	return d.vluDecode(64)
}

// vli (variable length integer) decoder
func (d *BytesDecoder) vliDecode(bits int) (value int64) {
	b := d.Uint8()
	sign := b & 0x40

	// first group of 6 bits
	value = int64(b & 0x3f)
	s := 6

	// while continuation bit is set
	for (b & 0x80) != 0 {
		if s >= bits {
			panic("integer representation too long")
		}

		// next group of 7 bits
		b = d.Uint8()
		value |= int64(b&0x7f) << s
		s += 7
	}

	if sign == 0 {
		// positive, sign bits are already zero
		return value
	}

	// negative, extend sign bits
	return value | (int64(-1) << s)
}

// vlu (variable length unsigned) decoder
func (d *BytesDecoder) vluDecode(bits int) uint64 {
	// first group of 7 bits
	b := d.Uint8()
	value := uint64(b & 0x7f)
	s := 7

	// while continuation bit is set
	for (b & 0x80) != 0 {
		if s >= bits {
			panic("integer representation too long")
		}

		// next group of 7 bits
		b = d.Uint8()
		value |= uint64(b&0x7f) << s
		s += 7
	}
	return value
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type BytesEncoder struct {
	data []byte
}

func NewBytesEncoder() *BytesEncoder {
	return &BytesEncoder{data: make([]byte, 0, 128)}
}

func (e *BytesEncoder) Address(value ScAddress) *BytesEncoder {
	return e.FixedBytes(value.Bytes(), uint32(TypeSizes[TYPE_ADDRESS]))
}

func (e *BytesEncoder) AgentID(value ScAgentID) *BytesEncoder {
	return e.FixedBytes(value.Bytes(), uint32(TypeSizes[TYPE_AGENT_ID]))
}

func (e *BytesEncoder) Bool(value bool) *BytesEncoder {
	if value {
		return e.Uint8(1)
	}
	return e.Uint8(0)
}

func (e *BytesEncoder) Bytes(value []byte) *BytesEncoder {
	length := uint32(len(value))
	e.Uint32(length)
	e.FixedBytes(value, length)
	return e
}

func (e *BytesEncoder) ChainID(value ScChainID) *BytesEncoder {
	return e.FixedBytes(value.Bytes(), uint32(TypeSizes[TYPE_CHAIN_ID]))
}

func (e *BytesEncoder) Color(value ScColor) *BytesEncoder {
	return e.FixedBytes(value.Bytes(), uint32(TypeSizes[TYPE_COLOR]))
}

func (e *BytesEncoder) Data() []byte {
	return e.data
}

func (e *BytesEncoder) FixedBytes(value []byte, length uint32) *BytesEncoder {
	if len(value) != int(length) {
		panic("invalid fixed bytes length")
	}
	e.data = append(e.data, value...)
	return e
}

func (e *BytesEncoder) Hash(value ScHash) *BytesEncoder {
	return e.FixedBytes(value.Bytes(), uint32(TypeSizes[TYPE_HASH]))
}

func (e *BytesEncoder) Hname(value ScHname) *BytesEncoder {
	return e.FixedBytes(value.Bytes(), uint32(TypeSizes[TYPE_HNAME]))
}

func (e *BytesEncoder) Int8(value int8) *BytesEncoder {
	return e.Uint8(uint8(value))
}

func (e *BytesEncoder) Int16(value int16) *BytesEncoder {
	return e.Int64(int64(value))
}

func (e *BytesEncoder) Int32(value int32) *BytesEncoder {
	return e.Int64(int64(value))
}

// vli (variable length integer) encoder
func (e *BytesEncoder) Int64(value int64) *BytesEncoder {
	// first group of 6 bits
	// 1st byte encodes 0 as positive in bit 6
	b := byte(value) & 0x3f
	value >>= 6

	finalValue := int64(0)
	if value < 0 {
		// encode negative value
		// 1st byte encodes 1 as negative in bit 6
		b |= 0x40
		finalValue = -1
	}

	// keep shifting until all bits are done
	for value != finalValue {
		// emit with continuation bit
		e.data = append(e.data, b|0x80)

		// next group of 7 bits
		b = byte(value) & 0x7f
		value >>= 7
	}

	// emit without continuation bit
	e.data = append(e.data, b)
	return e
}

func (e *BytesEncoder) RequestID(value ScRequestID) *BytesEncoder {
	return e.FixedBytes(value.Bytes(), uint32(TypeSizes[TYPE_REQUEST_ID]))
}

func (e *BytesEncoder) String(value string) *BytesEncoder {
	return e.Bytes([]byte(value))
}

func (e *BytesEncoder) Uint8(value uint8) *BytesEncoder {
	e.data = append(e.data, value)
	return e
}

func (e *BytesEncoder) Uint16(value uint16) *BytesEncoder {
	return e.Uint64(uint64(value))
}

func (e *BytesEncoder) Uint32(value uint32) *BytesEncoder {
	return e.Uint64(uint64(value))
}

// vlu (variable length unsigned) encoder
func (e *BytesEncoder) Uint64(value uint64) *BytesEncoder {
	// first group of 7 bits
	b := byte(value)
	value >>= 7

	// keep shifting until all bits are done
	for value != 0 {
		// emit with continuation bit
		e.data = append(e.data, b|0x80)

		// next group of 7 bits
		b = byte(value)
		value >>= 7
	}

	// emit without continuation bit
	e.data = append(e.data, b)
	return e
}
